src/infra/core/src/ingest/data_format_registry_impl.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use datafusion::prelude::SessionContext;
+use kamu_core::ingest::*;
+use kamu_ingest_datafusion::*;
+use opendatafabric::*;
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct DataFormatRegistryImpl {}
+
+#[dill::component(pub)]
+impl DataFormatRegistryImpl {
+    pub const FMT_CSV: DataFormatDesc = DataFormatDesc {
+        short_name: "CSV",
+        media_type: MediaType::CSV,
+    };
+    pub const FMT_JSON: DataFormatDesc = DataFormatDesc {
+        short_name: "JSON",
+        media_type: MediaType::JSON,
+    };
+    pub const FMT_NDJSON: DataFormatDesc = DataFormatDesc {
+        short_name: "NDJSON",
+        media_type: MediaType::NDJSON,
+    };
+    pub const FMT_GEOJSON: DataFormatDesc = DataFormatDesc {
+        short_name: "GeoJSON",
+        media_type: MediaType::GEOJSON,
+    };
+    pub const FMT_NDGEOJSON: DataFormatDesc = DataFormatDesc {
+        short_name: "NDGeoJSON",
+        media_type: MediaType::NDGEOJSON,
+    };
+    pub const FMT_PARQUET: DataFormatDesc = DataFormatDesc {
+        short_name: "Parquet",
+        media_type: MediaType::PARQUET,
+    };
+    pub const FMT_ESRI_SHAPEFILE: DataFormatDesc = DataFormatDesc {
+        short_name: "Shapefile",
+        media_type: MediaType::ESRI_SHAPEFILE,
+    };
+
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+#[async_trait::async_trait]
+impl DataFormatRegistry for DataFormatRegistryImpl {
+    fn list_formats(&self) -> Vec<DataFormatDesc> {
+        vec![
+            Self::FMT_CSV,
+            Self::FMT_JSON,
+            Self::FMT_NDJSON,
+            Self::FMT_GEOJSON,
+            Self::FMT_NDGEOJSON,
+            Self::FMT_PARQUET,
+            Self::FMT_ESRI_SHAPEFILE,
+        ]
+    }
+
+    fn format_of(&self, conf: &ReadStep) -> DataFormatDesc {
+        match conf {
+            ReadStep::Csv(_) => Self::FMT_CSV,
+            ReadStep::Json(_) => Self::FMT_JSON,
+            ReadStep::NdJson(_) => Self::FMT_NDJSON,
+            ReadStep::JsonLines(_) => Self::FMT_NDJSON,
+            ReadStep::GeoJson(_) => Self::FMT_GEOJSON,
+            ReadStep::NdGeoJson(_) => Self::FMT_NDGEOJSON,
+            ReadStep::Parquet(_) => Self::FMT_PARQUET,
+            ReadStep::EsriShapefile(_) => Self::FMT_ESRI_SHAPEFILE,
+        }
+    }
+
+    async fn get_reader(
+        &self,
+        ctx: SessionContext,
+        conf: ReadStep,
+        temp_path: PathBuf,
+    ) -> Result<Arc<dyn Reader>, ReadError> {
```

Comment:
Maybe the name `Reader` is a bit too general.

Reply:
I actually would like us to use modularization/namespacing more instead of long and complex type names, e.g.

Instead of: `kamu::DatasetIngestReader`
Use: `kamu::ingest::Reader`

I think this would create nice small sub-domain languages instead of really large and complex single domain language.

But I do see that without looking at imports name `Reader` can look super generic and confusing. Will try to come up with a nice middleground (but outside of this PR).